Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp

```diff
@@ -239,27 +239,57 @@ void Storage::WriteOperation::invokeCompletionHandler(int result)
         m_completionHandler(result);
 }
 
-struct Storage::TraverseOperation {
-    WTF_MAKE_TZONE_ALLOCATED(Storage::TraverseOperation);
+class TraverseOperation final : public ThreadSafeRefCounted<TraverseOperation, WTF::DestructionThread::MainRunLoop> {
 public:
-    TraverseOperation(Ref<Storage>&& storage, const String& type, OptionSet<TraverseFlag> flags, TraverseHandler&& handler)
-        : storage(WTFMove(storage))
-        , type(type)
-        , flags(flags)
-        , handler(WTFMove(handler))
-    { }
-    Ref<Storage> storage;
+    static Ref<TraverseOperation> create(Storage::TraverseHandler&& handler)
+    {
+        ASSERT(isMainRunLoop());
+        return adoptRef(*new TraverseOperation(WTFMove(handler)));
+    }
 
-    const String type;
-    const OptionSet<TraverseFlag> flags;
-    const TraverseHandler handler;
+    void invokeHandler(const Storage::Record* record, const Storage::RecordInfo& info)
+    {
+        ASSERT(isMainRunLoop());
+        m_handler(record, info);
+    }
 
-    Lock activeLock;
-    Condition activeCondition;
-    unsigned activeCount WTF_GUARDED_BY_LOCK(activeLock) { 0 };
-};
+    static const unsigned maxParallelActivityCount = 5;
+    void waitAndIncrementActivityCount()
+    {
+        Locker locker { m_lock };
+        m_activeCondition.wait(m_lock, [this] {
+            assertIsHeld(m_lock);
+            return m_activityCount < maxParallelActivityCount;
+        });
+        ++m_activityCount;
+    }
+
+    void decrementActivityCount()
+    {
+        Locker locker { m_lock };
+        --m_activityCount;
+        m_activeCondition.notifyOne();
+    }
+
+    void waitUntilActivitiesFinished()
+    {
+        Locker locker { m_lock };
+        m_activeCondition.wait(m_lock, [this] {
+            assertIsHeld(m_lock);
+            return !m_activityCount;
```

Comment:
I'm not sure how `waitUntilActivitiesFinished()` is called, but this seems potential problem:
1. thread A is active. `m_activityCount == 1` 
2. main thread is waiting by `waitUntilActivitiesFinished()`
3. thread A call `decrementActivityCount()`
4. during that, thread B call `waitAndIncrementActivityCount()`
5. thread A calls `notifyOne()` `m_activityCount == 0`
6. both main thread and thread B can be called, but if main thread is chosen, it satisfy the condition and goes on while thread B is still working.

Reply:
ah, the key is in existing usage, `waitAndIncrementActivityCount()` and `waitUntilActivitiesFinished()` is invoked in order on the same thread (in the same task dispatched to ` ioQueue()`). `decrementActivityCount()` is invoked on the main thread (IOChannel dispatches results back to main).